# Домашнее задание

In [1]:
import pandas as pd

## Задание 1
Скачайте с сайта https://grouplens.org/datasets/movielens/ датасет любого размера. Определите какому фильму было выставлено больше всего оценок 5.0.

In [6]:
data_1 = pd.read_csv('ml-latest-small/ratings.csv')
data_1.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [7]:
data_2 = pd.read_csv('ml-latest-small/movies.csv')
data_2.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [15]:
data = pd.merge(data_1, data_2, on = 'movieId', how = 'left')
data.head()

userId  movieId  rating  timestamp                        title  \
0       1        1     4.0  964982703             Toy Story (1995)   
1       1        3     4.0  964981247      Grumpier Old Men (1995)   
2       1        6     4.0  964982224                  Heat (1995)   
3       1       47     5.0  964983815  Seven (a.k.a. Se7en) (1995)   
4       1       50     5.0  964982931   Usual Suspects, The (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                               Comedy|Romance  
2                        Action|Crime|Thriller  
3                             Mystery|Thriller  
4                       Crime|Mystery|Thriller

In [24]:
filtered_data = data [data['rating'] == 5.0]
filtered_data.head()

userId  movieId  rating  timestamp                        title  \
3       1       47     5.0  964983815  Seven (a.k.a. Se7en) (1995)   
4       1       50     5.0  964982931   Usual Suspects, The (1995)   
6       1      101     5.0  964980868         Bottle Rocket (1996)   
8       1      151     5.0  964984041               Rob Roy (1995)   
9       1      157     5.0  964984100        Canadian Bacon (1995)   

                           genres  
3                Mystery|Thriller  
4          Crime|Mystery|Thriller  
6  Adventure|Comedy|Crime|Romance  
8        Action|Drama|Romance|War  
9                      Comedy|War

In [25]:
filtered_data['movieId'].value_counts().head()

318     153
296     123
356     116
2571    109
260     104
Name: movieId, dtype: int64

In [33]:
filtered_data[filtered_data['movieId'] == 318].head()

userId  movieId  rating   timestamp                             title  \
703        6      318     5.0   845553200  Shawshank Redemption, The (1994)   
1051       8      318     5.0   839463489  Shawshank Redemption, The (1994)   
1443      15      318     5.0  1510571866  Shawshank Redemption, The (1994)   
1677      17      318     5.0  1305696465  Shawshank Redemption, The (1994)   
1797      18      318     5.0  1455049328  Shawshank Redemption, The (1994)   

           genres  
703   Crime|Drama  
1051  Crime|Drama  
1443  Crime|Drama  
1677  Crime|Drama  
1797  Crime|Drama

### Ответ: 
фильм с наибольшим количеством оценок 5.0 (153 оценки) - Shawshank Redemption, The (1994), с movieId = 318

## Задача 2

По данным файла power.csv посчитайте суммарное потребление стран Прибалтики (Латвия, Литва и Эстония) категорий 4, 12 и 21 за период с 2005 по 2010 года. Не учитывайте в расчетах отрицательные значения quantity.

In [34]:
df = pd.read_csv('power.csv')
df.head()

country  year  quantity  category
0  Austria  1996       5.0         1
1  Austria  1995      17.0         1
2  Belgium  2014       0.0         1
3  Belgium  2013       0.0         1
4  Belgium  2012      35.0         1

In [47]:
baltic_df = df[(df['country'] == 'Lithuania') | (df['country'] == 'Latvia') | (df['country'] == 'Estonia')]
baltic_df.head()

country  year  quantity  category
373  Lithuania  2014      39.0         1
374  Lithuania  2013     192.0         1
375  Lithuania  2012     196.0         1
376  Lithuania  2011     171.0         1
377  Lithuania  2010     158.0         1

In [56]:
final_data = baltic_df[((baltic_df['category'] == 4) | (baltic_df['category'] == 12) | (baltic_df['category'] == 21)) & (baltic_df['year'] >= 2005) & (baltic_df['year'] <= 2010) & (baltic_df['quantity'] >= 0)]
final_data.head(100)

country  year  quantity  category
10617     Estonia  2008       1.0         4
10618     Estonia  2007       0.0         4
10619     Estonia  2006       0.0         4
10620     Estonia  2005       1.0         4
11156   Lithuania  2008       1.0         4
...           ...   ...       ...       ...
107661  Lithuania  2007     165.0        12
107662  Lithuania  2006       3.0        12
107663  Lithuania  2005       3.0        12
108496  Lithuania  2010       8.0        12
108497  Lithuania  2009      30.0        12

[100 rows x 4 columns]

In [58]:
final_data['quantity'].sum()

240580.0

## Задача 3
Выберите страницу любого сайта с табличными данными. Импортируйте таблицы в pandas dataframe.
Примеры страниц (необязательно брать именно эти):
https://fortrader.org/quotes
https://www.finanz.ru/valyuty/v-realnom-vremeni

In [71]:
tables = pd.read_html('https://fortrader.org/quotes')
tables[5]

Type           Last  \
0                               Bitcoin Курс биткоина    9086.700000   
1          Litecoin Курс лайткоин к доллару (LTC/USD)      59.142000   
2   BitcoinCash Курс Биткоин кеш к доллару США (BC...     366.880000   
3                                           Dash Dash     113.932000   
4                                       Ripple Ripple       0.237060   
5                                   Ethereum Ethereum     173.080000   
6                                   EmerCoin EmerCoin       0.140050   
7                                   NameCoin NameCoin       0.274000   
8                                   PeerCoin PeerCoin       1.300000   
9                             BTCRUB биткоин в рублях  247059.000000   
10                              BTCEUR биткоин в евро    8212.590000   
11                                        Zcash (ZEC)      56.049000   
12                                          NEO (NEO)      11.262000   
13                                          EOS (EOS)       3.999300   
14                                         TRON (TRX)       0.018531   
15                                             Monero      66.412000   
16                                            Cardano       0.051471   
17                                               IOTA       0.284040   
18                       ETC/USD эфириум классик, ETC      11.472000   

         Chg.      Chg.%  Time  
0    +145.200   1.624000   NaN  
1      +0.504   0.860000   NaN  
2      -2.570  -0.696000   NaN  
3      -0.075  -0.066000   NaN  
4    +0.00426   1.829900   NaN  
5      +1.950   1.139000   NaN  
6     0.00000   0.000000   NaN  
7      -0.051 -15.692000   NaN  
8       +0.16  14.040000   NaN  
9   -3 097.00  -1.240000   NaN  
10   +129.390   1.601000   NaN  
11     +0.838   1.518000   NaN  
12     +0.181   1.633000   NaN  
13    +0.0343   0.865100   NaN  
14  +0.001418   8.286098   NaN  
15     +0.773   1.178000   NaN  
16  +0.003559   7.428202   NaN  
17   +0.03138  12.419850   NaN  
18     +1.044  10.012000   NaN